README.md
# SLAnimation
Simple and powerful animation library built in Objective-C, compatible with Swift.


### Features:
* Numerous animatable properties
* Multi-stage animations
* Callbacks
* Simplified Effects Wrapper

## Installation

SLAnimation is available through [CocoaPods](http://cocoapods.org). To install
it, simply add the following line to your Podfile:

```ruby
pod 'SLAnimation'
```


# Class: SLAnimUtil
Common effects at your finger tips.

### Effects:
* Shake
* Pop (in/out)
* Rise (in/out)
* Fall (in/out)
* Slide Left (in/out)
* Slide Right (in/out)
* Bounce Left (in/out)
* Bounce Right (in/out)
* Unroll Down (in/out)
* Fade Down (in/out)
* Fade Up (in/out)

### Usage:
**Effect: Simple**
```objc
[SLAnimUtil animEffect:SLA_EFFECT_SHAKE view:view time:0.5];
```

**Effect: With Delay**
```objc
[SLAnimUtil animEffect:SLA_EFFECT_SHAKE view:view time:0.5 delay:0.5];
```

**Effect: With Delay + Callback**
```objc
[SLAnimUtil animEffect:SLA_EFFECT_SHAKE view:view time:0.5 delay:0.5 cbTarget:self cbSelector:@selector(onAnimateComplete)];
```

### Example:
**Shake**

![Gif](https://raw.githubusercontent.com/jclwong/SLAnimation/master/readme-assets/sl-effects-shake.gif "Shake")

**Pop (In/Out)**

![Gif](https://raw.githubusercontent.com/jclwong/SLAnimation/master/readme-assets/sl-effects-pop-in-out.gif "Pop In/Out")

**Rise (In/Out) -> No spring motion**

![Gif](https://raw.githubusercontent.com/jclwong/SLAnimation/master/readme-assets/sl-effects-rise-in-out.gif "Rise In/Out")

**Fall (In/Out) -> No spring motion**

![Gif](https://raw.githubusercontent.com/jclwong/SLAnimation/master/readme-assets/sl-effects-fall-in-out.gif "Fall In/Out")

**Slide Left (In/Out)**

![Gif](https://raw.githubusercontent.com/jclwong/SLAnimation/master/readme-assets/sl-effects-slide-left-in-out.gif "Slide Left In/Out")


**Bounce Left (In/Out) -> i.e. Slide with a slight spring motion**

![Gif](https://raw.githubusercontent.com/jclwong/SLAnimation/master/readme-assets/sl-effects-bounce-left-in-out.gif "Bounce Left In/Out")


**Unroll Down (In/Out)**

![Gif](https://raw.githubusercontent.com/jclwong/SLAnimation/master/readme-assets/sl-effects-unroll-down-in-out.gif "Unroll Down In/Out")

**Fade Down (In/Out) -> Small vertical displacement, with fade **

**Fade Up (In/Out) -> Small vertical displacement, with fade **


# Class: SLAnim
Core class for advanced usage


### Compulsary
* View
* Time

### Animatable Properties:
* Alpha
* Center
* Color: Background
* Frame
* Shadow Opacity
* Transform X, Transform Y, Transform XY

### Optional
* Delay
* Callback (+ Callback Object)


### Usage:
```objc
SLAnim *anim = [[SLAnim alloc] init];
// (Set Compulsary)
// (Set Animation Properties)
// (Set Optionals)
[anim animate];
```

#### Use-case: Pop In Effect
* Scale up (slightly larger)
* Scale down (correct size)
```objc
SLAnim *anim = [[SLAnim alloc] initWithView:view];
[anim setTimes:@[@(0.4), @(0.1)]];
[anim setProp:SLA_PROP_TRANSFORM_XY values:@[@0, @1.1, @1]];
[anim animate];
```

#### Use-case[2]: Pulse Alpha
* Fade In
* Fade Out Slow
* Optional: Trigger callbacks
```objc
SLAnim *anim = [[SLAnim alloc] initWithView:view];
[anim setTimes:@[@(0.1), @(0.4)]];
[anim setProp:SLA_PROP_ALPHA values:@[@0, @1, @0]];
[anim setCallbackTarget:self callbackSelector:@selector(onAnimateComplete)];
[anim animate];
```


# Author

Jack Wong

# License

SLAnimation is available under the MIT license. See the LICENSE file for more info.


CHANGELOG.md
All notable changes to this project will be documented in this file.

## [1.4.2] 2018-01-15
### Changed
- Animation properties to use "NSUInteger" instead of "int" for Swift compatibility.

## [1.4] 2017-12-19
### Added
- Added Effects: Fade Down
- Added Effects: Fade Up

## [1.3] 2017-11-25
### Added
- Added Animatable Property: Rotation
- Added CHANGELOG


## [1.2] 2017-10-23
### Added
- Added Effects: Rise
- Added Effects: Fall

### Changed
- Changed effect types from NSValue to NSObject - to hold colors properly

## [1.1] 2017-10-17
### Added
- Hello CocoaPods!


## [1.0] 2017-10-16
### Changed

- Both classes now have generic "setProp" or "setEffect" which uses ENUMS for clarity - instead of separate methods to set each prop.
- Implemented designated initialisers - now every effect has access to the same (full) functionality, ie callbacks, callback objects, etc.
- Refactored: SLAnimation to SLAnim
- Refactored: SLAnimationUtil to SLAnimUtil

## [0.3] 2017-11-25
### Added
- Added Effects: Slide Left
- Added Effects: Slide Right

## [0.2] 2017-11-25
### Added
- Added Effects: Pop
- Added Effects: Bounce Left
- Added Effects: Bounce Right
- Added Effects: Unroll

### Changed
- Changed Shake Effect to default without additional parameters for simplicity. If detailed control required, use SLAnim directly


## [0.1] 2017-5-31
### Added
- Added Animatable Property: Alpha
- Added Animatable Property: Background Color
- Added Animatable Property: Tint Color
- Added Animatable Property: Center
- Added Animatable Property: Frame
- Added Animatable Property: Shadow Opacity
- Added Animatable Property: Transform Scale
- Added Animatable Property: Transform Vertical Scale
- Added Timing curves
- Added Effects: Shake


SLAnimUtil.m
//
//  SLAnimUtil.m
//  StepLearn Education
//
//  Created by Jack Wong on 31/5/17.
//  Copyright © 2017 StepLearn Education. All rights reserved.
//

#import "SLAnimUtil.h"
#import "SLAnim.h"

@implementation SLAnimUtil

#pragma mark - Animations: Single Property

+ (BOOL) animProp:(int)prop animValue:(NSValue *)animValue view:(UIView *)view time:(float)time {
    return [self animProp:prop animValue:animValue view:view time:time delay:0 cbTarget:nil cbSelector:nil cbObject:nil];
}
+ (BOOL) animProp:(int)prop animValue:(NSValue *)animValue view:(UIView *)view time:(float)time delay:(float)delay {
    return [self animProp:prop animValue:animValue view:view time:time delay:delay cbTarget:nil cbSelector:nil cbObject:nil];
}
+ (BOOL) animProp:(int)prop animValue:(NSValue *)animValue view:(UIView *)view time:(float)time delay:(float)delay cbTarget:(id)target cbSelector:(SEL)selector {
    return [self animProp:prop animValue:animValue view:view time:time delay:delay cbTarget:target cbSelector:selector cbObject:nil];
}

+ (BOOL) animProp:(int)prop animValue:(NSValue *)animValue
             view:(UIView *)view time:(float)time delay:(float)delay
         cbTarget:(id)target cbSelector:(SEL)selector cbObject:(id)cbObject {
    SLAnim *anim = [[SLAnim alloc] initWithView:view times:@[@(time)]];
    [anim setDelay:delay];
    [anim setCallbackTarget:target callbackSelector:selector callbackObject:cbObject];
    
    if (prop == SLA_PROP_ALPHA) {
        [anim setProp:SLA_PROP_ALPHA values:@[@(view.alpha), animValue]];

    } else if (prop == SLA_PROP_CENTER) {
        [anim setProp:SLA_PROP_CENTER values:@[[NSValue valueWithCGPoint:view.center], animValue]];
        
    } else if (prop == SLA_PROP_COLOR_BG) {
        [anim setProp:SLA_PROP_COLOR_BG values:@[view.backgroundColor, animValue]];

    } else if (prop == SLA_PROP_COLOR_TINT) {
        [anim setProp:SLA_PROP_COLOR_TINT values:@[view.tintColor, animValue]];

    } else if (prop == SLA_PROP_FRAME) {
        [anim setProp:SLA_PROP_FRAME values:@[[NSValue valueWithCGRect:view.frame], animValue]];

    } else if (prop == SLA_PROP_SHADOW_OPACITY) {
        [anim setProp:SLA_PROP_SHADOW_OPACITY values:@[@(view.layer.shadowOpacity), animValue]];
    
    } else if (prop == SLA_PROP_TRANSFORM_X) {
        [anim setProp:SLA_PROP_TRANSFORM_X values:@[@(view.transform.a), animValue]];
        
    } else if (prop == SLA_PROP_TRANSFORM_Y) {
        [anim setProp:SLA_PROP_TRANSFORM_Y values:@[@(view.transform.d), animValue]];
        
    } else if (prop == SLA_PROP_TRANSFORM_XY) {
        [anim setProp:SLA_PROP_TRANSFORM_XY values:@[@(view.transform.a), animValue]];   // a(x) == d(y)
    }
    return [anim animate];
}


#pragma mark - Animation: Effects

+ (BOOL) animEffect:(int)effect view:(UIView *)view time:(float)time {
    return [self animEffect:effect view:view time:time delay:0 cbTarget:nil cbSelector:nil cbObject:nil];
}
+ (BOOL) animEffect:(int)effect view:(UIView *)view time:(float)time delay:(float)delay {
    return [self animEffect:effect view:view time:time delay:delay cbTarget:nil cbSelector:nil cbObject:nil];
}
+ (BOOL) animEffect:(int)effect view:(UIView *)view time:(float)time delay:(float)delay cbTarget:(id)target cbSelector:(SEL)selector {
    return [self animEffect:effect view:view time:time delay:delay cbTarget:target cbSelector:selector cbObject:nil];
}
+ (BOOL) animEffect:(int)effect view:(UIView *)view time:(float)time delay:(float)delay cbTarget:(id)target cbSelector:(SEL)selector cbObject:(id)cbObject {
    SLAnim *anim = [[SLAnim alloc] initWithView:view];
    [anim setDelay:delay];
    [anim setCallbackTarget:target callbackSelector:selector callbackObject:cbObject];
    
    if (effect == SLA_EFFECT_BOUNCE_LEFT_IN) {
        [self bounceLeftIn:anim view:view time:time];
    } else if (effect == SLA_EFFECT_BOUNCE_LEFT_OUT) {
        [self bounceLeftOut:anim view:view time:time];
        
    } else if (effect == SLA_EFFECT_BOUNCE_RIGHT_IN) {
        [self bounceRightIn:anim view:view time:time];
    } else if (effect == SLA_EFFECT_BOUNCE_RIGHT_OUT) {
        [self bounceRightOut:anim view:view time:time];

    } else if (effect == SLA_EFFECT_POP_IN) {
        [self popIn:anim view:view time:time];
    } else if (effect == SLA_EFFECT_POP_OUT) {
        [self popOut:anim view:view time:time];

    } else if (effect == SLA_EFFECT_SHAKE) {
        [self shake:anim view:view time:time];

    } else if (effect == SLA_EFFECT_SLIDE_LEFT_IN) {
        [self slideLeftIn:anim view:view time:time];
    } else if (effect == SLA_EFFECT_SLIDE_LEFT_OUT) {
        [self slideLeftOut:anim view:view time:time];

    } else if (effect == SLA_EFFECT_SLIDE_RIGHT_IN) {
        [self slideRightIn:anim view:view time:time];
    } else if (effect == SLA_EFFECT_SLIDE_RIGHT_OUT) {
        [self slideRightOut:anim view:view time:time];

    } else if (effect == SLA_EFFECT_UNROLL_DOWN_IN) {
        [self unrollDownIn:anim view:view time:time];
    } else if (effect == SLA_EFFECT_UNROLL_DOWN_OUT) {
        [self unrollDownOut:anim view:view time:time];
    }
    return [anim animate];
}


#pragma mark - Shake

+ (SLAnim *) shake:(SLAnim *)anim view:(UIView *)view time:(float)time {
    [anim setTimes:@[@(time*0.2), @(time*0.2), @(time*0.2), @(time*0.2), @(time*0.2)]];
    [anim setProp:SLA_PROP_CENTER values:@[[NSValue valueWithCGPoint:view.center],
                                           [NSValue valueWithCGPoint:CGPointMake(view.center.x-15, view.center.y)],
                                           [NSValue valueWithCGPoint:CGPointMake(view.center.x+15, view.center.y)],
                                           [NSValue valueWithCGPoint:CGPointMake(view.center.x-10, view.center.y)],
                                           [NSValue valueWithCGPoint:CGPointMake(view.center.x+5, view.center.y)],
                                           [NSValue valueWithCGPoint:view.center]]];
    return anim;
}

#pragma mark - Pop

+ (SLAnim *) popIn:(SLAnim *)anim view:(UIView *)view time:(float)time {
    [anim setTimes:@[@(time*0.8), @(time*0.2)]];
    [anim setProp:SLA_PROP_ALPHA values:@[@0, @1, @1]];
    [anim setProp:SLA_PROP_TRANSFORM_XY values:@[@0.3, @1.1, @1]];
    return anim;
}

+ (SLAnim *) popOut:(SLAnim *)anim view:(UIView *)view time:(float)time {
    [anim setTimes:@[@(time*0.2), @(time*0.8)]];
    [anim setProp:SLA_PROP_ALPHA values:@[@1, @1, @0]];
    [anim setProp:SLA_PROP_TRANSFORM_XY values:@[@1, @1.1, @0.3]];
    return anim;
}


#pragma mark - Slide Left

+ (SLAnim *) slideLeftIn:(SLAnim *)anim view:(UIView *)view time:(float)time {
    float sWidth = [UIScreen mainScreen].bounds.size.width;
    [anim setTimes:@[@(time)]];
    [anim setAnimationOption:SLA_TIME_EASE_IN];
    [anim setProp:SLA_PROP_CENTER values:@[[NSValue valueWithCGPoint:CGPointMake(view.center.x+sWidth, view.center.y)], [NSValue valueWithCGPoint:view.center]]];
    [anim setProp:SLA_PROP_ALPHA values:@[@0, @1]];
    return anim;
}

+ (SLAnim *) slideLeftOut:(SLAnim *)anim view:(UIView *)view time:(float)time {
    float sWidth = [UIScreen mainScreen].bounds.size.width;
    [anim setTimes:@[@(time)]];
    [anim setAnimationOption:SLA_TIME_EASE_OUT];
    [anim setProp:SLA_PROP_CENTER values:@[[NSValue valueWithCGPoint:view.center], [NSValue valueWithCGPoint:CGPointMake(view.center.x - sWidth, view.center.y)]]];
    [anim setProp:SLA_PROP_ALPHA values:@[@1, @0]];
    return anim;
}

#pragma mark - Slide Right


+ (SLAnim *) slideRightIn:(SLAnim *)anim view:(UIView *)view time:(float)time {
    float sWidth = [UIScreen mainScreen].bounds.size.width;
    [anim setTimes:@[@(time)]];
    [anim setAnimationOption:SLA_TIME_EASE_IN];
    [anim setProp:SLA_PROP_CENTER values:@[[NSValue valueWithCGPoint:CGPointMake(view.center.x-sWidth, view.center.y)],
     [NSValue valueWithCGPoint:view.center]]];
    [anim setProp:SLA_PROP_ALPHA values:@[@0, @1]];
    return anim;
}

+ (SLAnim *) slideRightOut:(SLAnim *)anim view:(UIView *)view time:(float)time {
    float sWidth = [UIScreen mainScreen].bounds.size.width;
    [anim setTimes:@[@(time)]];
    [anim setAnimationOption:SLA_TIME_EASE_OUT];
    [anim setProp:SLA_PROP_CENTER values:@[[NSValue valueWithCGPoint:view.center],
     [NSValue valueWithCGPoint:CGPointMake(view.center.x+sWidth, view.center.y)]]];
    [anim setProp:SLA_PROP_ALPHA values:@[@1, @0]];
    return anim;
}

#pragma mark - Bounce Left

+ (SLAnim *) bounceLeftIn:(SLAnim *)anim view:(UIView *)view time:(float)time {
    float sWidth = [UIScreen mainScreen].bounds.size.width;
    [anim setTimes:@[@(time*0.8), @(time*0.2)]];
    [anim setProp:SLA_PROP_ALPHA values:@[[NSValue valueWithCGPoint:CGPointMake(view.center.x+sWidth, view.center.y)],
     [NSValue valueWithCGPoint:CGPointMake(view.center.x-20, view.center.y)],
     [NSValue valueWithCGPoint:view.center]]];
    [anim setProp:SLA_PROP_ALPHA values:@[@0, @1, @1]];
    return anim;
}

+ (SLAnim *) bounceLeftOut:(SLAnim *)anim view:(UIView *)view time:(float)time {
    float sWidth = [UIScreen mainScreen].bounds.size.width;
    [anim setTimes:@[@(time*0.2), @(time*0.8)]];
    [anim setProp:SLA_PROP_CENTER values:@[[NSValue valueWithCGPoint:view.center],
     [NSValue valueWithCGPoint:CGPointMake(view.center.x+20, view.center.y)],
     [NSValue valueWithCGPoint:CGPointMake(view.center.x-sWidth, view.center.y)]]];
    [anim setProp:SLA_PROP_ALPHA values:@[@1, @1, @0]];
    return anim;
}

#pragma mark - Bounce Right

+ (SLAnim *) bounceRightIn:(SLAnim *)anim view:(UIView *)view time:(float)time {
    float sWidth = [UIScreen mainScreen].bounds.size.width;
    [anim setTimes:@[@(time*0.8), @(time*0.2)]];
    [anim setProp:SLA_PROP_CENTER values:@[[NSValue valueWithCGPoint:CGPointMake(view.center.x-sWidth, view.center.y)],
     [NSValue valueWithCGPoint:CGPointMake(view.center.x+20, view.center.y)],
     [NSValue valueWithCGPoint:view.center]]];
    [anim setProp:SLA_PROP_ALPHA values:@[@0, @1, @1]];
    return anim;
}

+ (SLAnim *) bounceRightOut:(SLAnim *)anim view:(UIView *)view time:(float)time {
    float sWidth = [UIScreen mainScreen].bounds.size.width;
    [anim setTimes:@[@(time*0.2), @(time*0.8)]];
    [anim setProp:SLA_PROP_CENTER values:@[[NSValue valueWithCGPoint:view.center],
     [NSValue valueWithCGPoint:CGPointMake(view.center.x-20, view.center.y)],
     [NSValue valueWithCGPoint:CGPointMake(view.center.x+sWidth, view.center.y)]]];
    [anim setProp:SLA_PROP_ALPHA values:@[@1, @1, @0]];
    return anim;
}

#pragma mark - Unroll Down

+ (SLAnim *) unrollDownIn:(SLAnim *)anim view:(UIView *)view time:(float)time {
    CGRect f = view.frame;
    [anim setTimes:@[@(time*0.8), @(time*0.2)]];
    [anim setProp:SLA_PROP_ALPHA values:@[@0, @1, @1]];
    [anim setProp:SLA_PROP_FRAME values: @[[NSValue valueWithCGRect:CGRectMake(f.origin.x, f.origin.y, f.size.width, 0)],
                     [NSValue valueWithCGRect:CGRectMake(f.origin.x, f.origin.y, f.size.width, f.size.height + f.size.height * 0.05)],
                     [NSValue valueWithCGRect:f]]];
    return anim;
}

+ (SLAnim *) unrollDownOut:(SLAnim *)anim view:(UIView *)view time:(float)time {
    CGRect f = view.frame;
    [anim setTimes:@[@(time*0.2), @(time*0.8)]];
    [anim setProp:SLA_PROP_ALPHA values:@[@1, @1, @0]];
    [anim setProp:SLA_PROP_FRAME values: @[[NSValue valueWithCGRect:f],
                     [NSValue valueWithCGRect:CGRectMake(f.origin.x, f.origin.y, f.size.width, f.size.height + f.size.height * 0.05)],
                     [NSValue valueWithCGRect:CGRectMake(f.origin.x, f.origin.y, f.size.width, 0)]]];
    return anim;
}


@end
